### PR TITLE
bpo-46649: Propagate Python thread name to PyThreadState

### PR DESCRIPTION
**Pull request by P403n1x87**
This change allows Python thread objects to propagate their name to the underlying PyThreadState structure to make it easier for tools that rely on the ABI, like sampling profilers, to retrieve this information and provide richer profiling data.

https://bugs.python.org/issue46649

### CODE REVIEW

File: Modules/_threadmodule.c [hunk 1]
```diff
@@ -1551,6 +1551,40 @@ PyDoc_STRVAR(excepthook_doc,
 \n\
 Handle uncaught Thread.run() exception.");
 
+
+static PyObject*
+thread_set_name(PyObject *self, PyObject *args) {
```
Comment (JelleZijlstra) [inline Modules/_threadmodule.c:1595]:
```
 % ./python.exe
Python 3.11.0a5+ (heads/bpo-412569/propagate-thread-name:9f9acf49be, Feb  5 2022, 17:15:00) [Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import _thread
>>> _thread.set_name(123434, "x")
zsh: segmentation fault  ./python.exe
```

This function basically just writes to arbitrary memory, which is a security risk. I think the right fix would be to pass the thread state around as some sort of opaque object instead of a raw int.

File: Lib/test/test_threading.py [hunk 2]
```diff
@@ -961,6 +962,37 @@ def import_threading():
...
+        thread = threading.Thread(target=target, name="test-thread-name")
+        thread.start()
+
+        main_thread = threading.current_thread()
+        self.assertIs(main_thread.name, _testcapi.get_thread_name(main_thread._tstate))
+
+        thread.join()
+
+        # Check we can still set the name after the thread terminated
+        thread.name = "foo"
```
Comment (JelleZijlstra) [inline Lib/test/test_threading.py:1002]: This could use some more tests checking that the name you set is actually reflected.

Comment (P403n1x87) [reply Lib/test/test_threading.py:1002]: This was meant to catch any potential segfaults, since once the thread terminates there isn't a thread state anymore.

File: Modules/_threadmodule.c (hunk 1 above)
Comment (P403n1x87) [reply Modules/_threadmodule.c:1595]: Good point. `set_name` is not meant to be called by anything else other than `Thread`, but it must be made safe nonetheless! Is there an established pattern for this sort of issue? I'm thinking of making an object, like `_PyThreadState_Reference` to pass from C to Python.

File: Include/pystate.h [hunk 3]
```diff
@@ -14,13 +14,20 @@ removed (with effort). */
...
 /* struct _ts is defined in cpython/pystate.h */
 typedef struct _ts PyThreadState;
+/* struct _tsr is defined in cpython/pystate.h */
+typedef struct _tsr PyTStateRef;
 /* struct _is is defined in internal/pycore_interp.h */
 typedef struct _is PyInterpreterState;
 
+PyAPI_DATA(PyTypeObject) PyTStateRef_Type;
+
+#define PyTStateRef_Check(op) Py_IS_TYPE(op, &PyTStateRef_Type)
```
Comment (JelleZijlstra) [inline Include/pystate.h:29]: I don't think any of this needs to go in a public header file.

File: Modules/_testcapimodule.c [hunk 4]
```diff
@@ -5770,6 +5770,13 @@ test_tstate_capi(PyObject *self, PyObject *Py_UNUSED(args))
     Py_RETURN_NONE;
 }
 
+static PyObject *
+get_thread_name(PyObject *self, PyTStateRef *tsr)
+{
+    assert(PyTStateRef_Check(tsr));
```
Comment (JelleZijlstra) [inline Modules/_testcapimodule.c]: Should raise an exception here instead of just asserting (similar to type_get_version above).

File: Modules/_threadmodule.c [hunk 5]
```diff
@@ -1551,6 +1551,102 @@ PyDoc_STRVAR(excepthook_doc,
...
+    return PyUnicode_FromFormat("<tstate reference object at %p>", tsr->tstate);
+}
+
+
+PyTypeObject PyTStateRef_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    "TStateRef",
+    sizeof(PyTStateRef),
+    0,
+    (destructor)PyObject_Free,          /* tp_dealloc */
```
Comment (JelleZijlstra) [inline Modules/_threadmodule.c]: I believe you can use designated initializers (`.tp_dealloc = `) to avoid all those 0s.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
What happens if you hold on to a `PyTStateRef` object after the corresponding thread has ended? I wasn't able to get a segfault this way, but it looks like I'd be writing into freed memory.

File: Include/pystate.h (hunk 3 above)
Comment (P403n1x87) [reply Include/pystate.h:29]: I've put this here so that `PyTStateRef_Check(op)` could be used in `_testcapi`. But given how simple the definition is I could get rid of it.

**Comment by P403n1x87** [comment]
> What happens if you hold on to a `PyTStateRef` object after the corresponding thread has ended? I wasn't able to get a segfault this way, but it looks like I'd be writing into freed memory.

Hmm yes I think it's still possible to get a segfault this way ☹️. One solution that I can think of is to cache the object on the PyThreadState instance, but it doesn't feel like a very clean approach to me 🤔

File: Include/pystate.h (hunk 3 above)
Comment (P403n1x87) [reply Include/pystate.h:29]: @JelleZijlstra I can't get this to work if I remove the `PyAPI_DATA` export. Any idea where I could put that to avoid affecting the stable API? I think one thing to consider, though, is that the new `PyTStateRef` objects are surfaced to Python.

Comment (JelleZijlstra) [reply Include/pystate.h:29]: I'm not too experienced with the C API, but I am pretty confident we don't want to expose this class in the C API without a strong reason. Generally we're trying to restrict the C API so that it is easier to evolve CPython. For example, changes to the PyThreadState have caused trouble several times. https://devguide.python.org/c-api/ says internal symbols should ideally be in Include/internal.

https://docs.python.org/3/c-api/init.html#c.PyThreadState also says that only the "interp" member of this struct is public, so putting it here isn't enough to allow well-behaved external users of the API to access the thread name. I think you should prefix the new struct members with underscores and add a new, documented and public function `PyThreadState_GetName()` for extensions to access the thread name.

File: Python/pystate.c [hunk 6]
```diff
@@ -783,7 +783,7 @@ init_threadstate(PyThreadState *tstate,
     tstate->datastack_chunk = NULL;
     tstate->datastack_top = NULL;
     tstate->datastack_limit = NULL;
-
+
```
Comment (JelleZijlstra) [inline Python/pystate.c:786]: unnecessary change

File: Python/pystate.c [hunk 7]
```diff
@@ -1034,6 +1034,11 @@ PyThreadState_Clear(PyThreadState *tstate)
     Py_CLEAR(tstate->async_gen_finalizer);
 
     Py_CLEAR(tstate->context);
+    
+    if (tstate->ref != NULL)
```
Comment (JelleZijlstra) [inline Python/pystate.c:1038]: I'm not sure how PyThreadState interacts with the garbage collector. How will the GC know that the thread state ref object is alive?

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks! Here's a few more comments.

File: Python/pystate.c (hunk 7 above)
Comment (P403n1x87) [reply Python/pystate.c:1038]: The ref object has a life on its own. The PyThreadState creates the ref object. Once the thread state is deallocated, it sets the actual ref inside the ref object to NULL and decrements its refcount. The ref object then lives on until there are references to it. Otherwise, it is GC'd at the next collection cycle.

**Review by pablogsal — CHANGES_REQUESTED** [review]
Unfortunately, I am not convinced this is a good idea. We are moving more and more to a private thread state, and we want to be able to modify its ABI, fields and ordering even between patch releases so I am not comfortable assuring that even more fields can be retrieved via a constant offset from the thread state. Also, adding Python objects to the thread state is possible, but makes destruction more tricky (for example, destroying Python objects must be done with the GIL held, but when the thread state is being teared down from an extension module the GIL may be unlocked).

Additionally, I think this PR is adding quite a lot of extra complexity for very restricted gains.

I am not strongly against this, but unless there is strong consensus that this is clearly a win, I am not confident with this.

**Comment by iritkatriel** [comment]
Closing since there were objections to the change and this has been idle for over a year since then.
